Project 3/Task 1/block_quantize.py

```diff
@@ -0,0 +1,126 @@
+import sys
+import math
+import numpy as np
+import cv2
+import matplotlib.pyplot as plt
+import re
+import Utility as util
+from os import listdir,path
+
+'''
+Quantizes a block using the number of bins
+Returns a formatted string for output to a file which details the information in the format:
+< FrameNumber, (X coordinate, Y coordinate) , Y Channel Value, Number Of Occurances >
+'''
+def quantizeBlock(block, bins, frameNum, x, y):
+    minValue = np.amin(block)
+    maxValue = np.amax(block)
+    samples, valuesPerBin = np.linspace(minValue, maxValue, bins, endpoint=False, retstep=True)
+    samples += 0.5*valuesPerBin
+    samples = map((lambda v: round(v, 7)), samples)
+    occurancesInFrame = dict.fromkeys(samples, 0) #dictionary to contain how many times a value occurs
+
+    for val in np.nditer(block):
+        if valuesPerBin == 0:
+            bottomBin = minValue
+        else:
+            bottomBin = minValue + (math.floor((val - minValue) / valuesPerBin) + 0.5) * valuesPerBin
+
+        if bottomBin >= maxValue:
+            bottomBin = minValue + (bins - 0.5)*valuesPerBin
+
+        bottomBin = round(bottomBin, 7)
+        occurancesInFrame[bottomBin] += 1
+
+    result = list()
+
+    for key in occurancesInFrame:
+        result.append({
+            'frameNum': frameNum,
+            'blockCoords': (x, y),
+            'key': key,
+            'val': occurancesInFrame[key]
+        })
+
+    return result
+
+def quantize(frameData, n):
+    print('Running quantization with ' + str(n) + ' bins...')
+
+    result = list()
+    frameNum = 0
+
+    #for each entry in frameData
+    for frame in frameData:
+        frameNum += 1
+        print('Processing regions of frame: ' + str(frameNum))
+        for blockI in range(0, len(frame), 8):
+            for blockJ in range(0, len(frame[blockI]), 8):
+                block = frame[blockI:blockI+8,blockJ:blockJ+8]
+                result.extend(quantizeBlock(block, n, frameNum, blockI, blockJ))
```

Comment:
If you want to leave it that's fine, but take a look at the report, not sure of the LOE of going back to redo all that.